### ISSUE TITLE
Rounded corners

### ISSUE THREAD

**Issue by ramnes** [qtile/qtile#1515]
```
03h56 aircraftblues ok so I'm new and have been playing around with the config and widgets. does anybody know where i can find info for rounded window corners?
```

See [here](https://www.reddit.com/r/unixporn/comments/benebi/i3gaps_i3_rounded_corners/) for an example and links to different implementations in bspwm and i3.

Definitely some great rice I'd be glad to see in Qtile.

**Comment by pawamoy** [comment]
(added a [draft](https://github.com/qtile/qtile/wiki/rices) of a "Rices" page in the wiki, would be great to show some)

**Comment by m-col** [comment]
I have a functioning [round corners commit](https://github.com/m-col/qtile/tree/round_corners) that I've used for a couple months. I haven't made a PR for it because it doesn't have anti-aliasing and I don't really know how to add it. Perhaps the composite extension in xcffib could be used but that guess is as far as I've got. This could serve as a starting point for rounded corners :)

**Comment by ramnes** [comment]
Awesome! I think it wouldn't hurt to get that into the codebase and then open another issue for the anti-aliasing, what do you think?

**Comment by m-col** [comment]
It's funny this is brought up now... I've recently been thinking I should open an issue for advice regarding anti-aliasing. I've made the PR #1517

**Comment by ramnes** [comment]
FWIW, it seems that picom can round the windows too, and that it works quite well with Qtile: https://www.reddit.com/r/UsabilityPorn/comments/k8s8ph/qtile_i_showed_off_my_bspwm_rice_earlier_so_heres/

Do you see any advantage of implementing this feature here rather than letting the compositor manage it?

**Comment by m-col** [comment]
Yep - picom cuts into borders, whereas as we have control over the borders we
can draw borders that round on both the inside and outside edge of those
borders.

**Comment by ramnes** [comment]
Are you sure? The borders aren't cut in the screenshot of Reddit post I mentioned, and picom doesn't seem to do borders, nor that fork.

**Comment by m-col** [comment]
Oh yeah, it does. Interesting. Arch repo's version of picom (which will be yushi's last release) doesn't seem to have this feature in any form yet, and in that PR you link the [only image](https://github.com/yshui/picom/pull/229#issuecomment-609532144) with rounded borders is a request for that behaviour. I'm assuming that the rounded borders are only part of jonaburg's fork (the one linked from the reddit post). They must be drawing their own borders over the top of the window('s borders provided by the WM). If that works people (and ideally is merged into yushi's fork) then that's great, but has potential caveats:

- does it know what colour the border should be? I suppose it might be able to somehow extract it from the colour configured from the WM.
- requires picom which in my experience is buggy as hell (to the extent where I recently just stopped using it completely).

Either way, if the picom fork works for you then all good! I am also happy to finish this PR off (just the tests?). The reason for it going dormant was that it conflicts massively with the multiple borders PR, which I have been pushing more (and is done, just needs merging).

**Comment by hekatonkheirex** [comment]
Hi all, I'd already made rounded corners with bspwm and i3, and using Polybar as a panel with the [Ibhagwan's fork](https://github.com/ibhagwan/picom). I can make an exclusion to not draw the panel borders with:
 `rounded-corners-exclude = [
"window_type = 'dock'",
"window_type = 'bar'",
"class_g = 'Polybar'",
"class_g = 'dmenu'",
`
But I can't make it work with the Qtile bar. Doing xprop to the bar I got:
`_NET_WM_WINDOW_OPACITY(CARDINAL) = 4294967295
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
QTILE_INTERNAL(CARDINAL) = 1`

So, only using `"window_type = "normal'"` seems to work, but that 'cancels' all the normal windows too.

**Comment by ivanddm** [comment]
@hekatonkheirex I have this in my picom config and it works for me:

```
rounded-corners-exclude = [
    "class_g = 'polybar'",
    "class_g = 'dmenu'",
    "class_g = 'awesome'",
    "window_type = 'dock'",
    "! name~=''",    # exclude windows with no name such as qtile's bar
];
```

**Comment by m-col** [comment]
You could use also 'QTILE_INTERNAL' directly, or perhaps any windows with a big
width and short height.

On 210323 11:50, Ivan Duran wrote:
***@***.*** I have this in my picom config and it works for me:

**Comment by hekatonkheirex** [comment]
> @hekatonkheirex I have this in my picom config and it works for me:
> 
> ```
> rounded-corners-exclude = [
>     "class_g = 'polybar'",
>     "class_g = 'dmenu'",
>     "class_g = 'awesome'",
>     "window_type = 'dock'",
>     "! name~=''",    # exclude windows with no name such as qtile's bar
> ];
> ```

Can confirm that this actually works! Thank you!

@m-col Already tried adding `'QTILE_INTERNAL'`, but it just behaves the same way having `"window_type = 'normal'",`.

**Comment by Barbarossa93** [comment]
I had a hard time with this as well, but for shadows on the bar. You have to treat the qtile bar as a cardinal window, so the rule you'd want to put in your exclude rules is:
```
"QTILE_INTERNAL:32c = 1"
```

**Comment by ryynison** [comment]
This might be a dumb question, but how would you exclude windows of certain sizes in picom? Such as the qtile bar or any other window?

> You could use also 'QTILE_INTERNAL' directly, or perhaps any windows with a big width and short height.

**Comment by Mister7F** [comment]
Hi :) I just waiting this feature to switch to qtile, I see that the PR is closed ( https://github.com/qtile/qtile/pull/1517 ), what's the status of this improvement ?

**Comment by m-col** [comment]
Hi @Mister7F, I wasn't super happy with how the implementation in the PR worked as it was kinda hacky. That said, I don't think I've seen rounded corners in a non-reparenting WM that were not hacky. If somebody were to continue working on it then I think we'd be happy to review and merge it. Since that PR, we now have a Wayland backend and support multiple borders, so these would both need taking into account too. I don't think we need to support using multiple corners and round corners are the same time though.
